simpletest/simpletest

```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+#: exec_target = cli
+
+## Run automated tests with SimpleTest.
+##
+## Usage: fin simpletest <args>
+
+cd "${PROJECT_ROOT}/${DOCROOT}"
+
+# Find the run-tests script.
+if [ -f "core/scripts/run-tests.sh" ]; then
```

Comment:
What is the reasoning behind this path? Is it canonical?

Reply:
This is to check if it's a Drupal 8 site, as the path to the script is different in D7 to D8. Unless there's a better way to do this?

Reply:
Ok I see. So it is kind of canonical path as per https://www.drupal.org/docs/8/phpunit/running-tests-through-command-line-with-run-testssh

So I would recommend covering all versions + giving ability to set this via variable.

```bash
#!/usr/bin/env bash

#: exec_target = cli

## Run automated tests with SimpleTest.
##
## Usage: fin simpletest <args>
##
## Options:
##   Set SIMPLETEST_PATH variable in docksal.env to override path to the SimpleTest script.
##   SIMPLETEST_PATH should be relative to the project root


DRUPAL8_PATH="core/scripts/run-tests.sh"
DRUPAL7_PATH="scripts/run-tests.sh"

# -- Runtime --

cd "${PROJECT_ROOT}/${DOCROOT}"

# If path to SIMPLETEST_PATH is set and exists
if [[ "$SIMPLETEST_PATH" != "" ]] && [[ -f "${PROJECT_ROOT}/$SIMPLETEST_PATH" ]]; then
  SCRIPT_FILE="${PROJECT_ROOT}/$SIMPLETEST_PATH"
# check for Drupal 8 path
elif [[ -f "$DRUPAL8_PATH" ]]; then
  SCRIPT_FILE="$DRUPAL8_PATH"
# check for Drupal 7 path
elif [[ -f "$DRUPAL7_PATH" ]]; then
  SCRIPT_FILE="$DRUPAL7_PATH"
else
  echo "ERROR: Could not find path to simpletest script."
  exit 1
fi

# Check that script is executable
chmod +x "$SCRIPT_FILE"

# Execute the script.
php "$SCRIPT_FILE" --color "$@"
```

Reply:
I'm not sure what the benefit is of making it configurable - it's always going to be in one place of D7 and another for D8.

Adding `DRUPAL8_PATH` and `DRUPAL7_PATH` variables does make it more readable, but if it needs to be that obvious, this could better be done with comments.

I have it working on a multi-site environment, which I'm using the `--url` option to determine which site to use. Had troubles making it work relatively before, though it would be good to make it as consistent as possible with the `phpunit` add-on.

Reply:
Please explain what prevents me from putting the script to any random folder and running it from there?

Reply:
I guess we can agree to disagree on the variable/comment thing.

To me, it's perfectly readable as it is with a clarifying comment then it is to create two extra variables - the only purpose of which is to assign one to be the value of another variable later in the code.

Without variables:

```
# Determine the correct path for the run-tests.sh script. This will depend on
# the version of Drupal.
if [ -f "core/scripts/run-tests.sh" ]; then
    SCRIPT_FILE="$PROJECT_ROOT/$DOCROOT/core/scripts/run-tests.sh"
else
    SCRIPT_FILE="$PROJECT_ROOT/$DOCROOT/scripts/run-tests.sh"
fi
```

With variables:

```
DRUPAL_8_PATH="$PROJECT_ROOT/$DOCROOT/core/scripts/run-tests.sh"
DRUPAL_7_PATH="$PROJECT_ROOT/$DOCROOT/scripts/run-tests.sh"

if [ -f DRUPAL_8_PATH ]; then
    SCRIPT_FILE="$DRUPAL_8_PATH"
else
    SCRIPT_FILE="$DRUPAL_7_PATH"
fi
```

Also, I'm not sure how I feel about being linked to the Clean Code book. It seems a little degrading, I'd prefer it if you could give feedback more constructively.

In terms of the random folder, in my current setup within my project root, `fin simpletest --directory sites/siteone/modules/custom --url http://siteone.docksal` works. Removing `--url` returns no output, and running it within `sites/siteone` as `fin simpletest --directory modules/custom --url http://siteone.docksal` only returns `ERROR: No valid tests were specified`.

This is the same as if I `fin bash`-ed into the container and ran `php ../../scripts/run-tests.sh --directory modules/custom --url http://siteone.docksal` I need to run it from `/var/www/web` and specify the full path to the directory.

You could do this:

```
php $SCRIPT_FILE \
  --php `which php` \
  --directory `pwd` \
  --color \
  "$@"
```

This works and runs the tests within the current directory (and sub-directories), but it's whether the `--directory` option should be set by default or if that's making it too opinionated?

Reply:
There is no canonical path or name for simple test scripts required by Simple Test. 

What if my simpletest script is `scripts/test/simpletest.sh`? How do I make this addon to run it?

Reply:
This code does not explain which version of Drupal is where:

```bash
# Determine the correct path for the run-tests.sh script. This will depend on
# the version of Drupal.
if [ -f "core/scripts/run-tests.sh" ]; then
    SCRIPT_FILE="$PROJECT_ROOT/$DOCROOT/core/scripts/run-tests.sh"
else
    SCRIPT_FILE="$PROJECT_ROOT/$DOCROOT/scripts/run-tests.sh"
fi
```


If you add documentation via comments it will actually be longer than using variables and less readable that using variables:

```bash
# Determine the correct path for the run-tests.sh script. This will depend on
# the version of Drupal.
if [ -f "core/scripts/run-tests.sh" ]; then
    # Drupal 8
    SCRIPT_FILE="$PROJECT_ROOT/$DOCROOT/core/scripts/run-tests.sh"
else
    # Drupal 7
    SCRIPT_FILE="$PROJECT_ROOT/$DOCROOT/scripts/run-tests.sh"
fi
```

Additionally, I have an experience of supporting 6000+ lines bash utility, and I know for a fact that it is better to use constants to define static paths and use those constants in runtime, because it is easier to maintain in the long run when you (or especially someone else) goes back to the code. So while we can agree to disagree, it is not just one random opinion against another random opinion.

Reply:
> What if my simpletest script is not "$PROJECT_ROOT/$DOCROOT/scripts/run-tests.sh" but <project root>/tests/run-tests.sh or <project root>/scripts/simpletest.sh? How do I make this addon to run it?

`run-test.sh` is part of core, so the paths will always be the same - `scripts/run-tests.sh` for D7 and `core/scripts/run-tests.sh` for D8. You wouldn't have `<project root>/scripts/simpletest.sh`.